### DISCUSSION TITLE
AIrflow scheduler crashes with a sqlalchemy.exc.InternalError: (psycopg2.errors.InFailedSqlTransaction)

### DISCUSSION THREAD

**Discussion by changxiaoju** [apache/airflow#20846] [Q&A]
when i run airflow scheduler in terminal, it always crash like this
```
[2022-01-12 21:58:58,969] {scheduler_job.py:596} INFO - 
Starting the scheduler
[2022-01-12 21:58:58,970] {scheduler_job.py:601} INFO - Processing each file at most -1 times
[2022-01-12 21:58:59 +0800] [23205] [INFO] Starting gunicorn 20.1.0
[2022-01-12 21:58:59 +0800] [23205] [ERROR] Connection in use: ('0.0.0.0', 9999)
[2022-01-12 21:58:59 +0800] [23205] [ERROR] Retrying in 1 second.
[2022-01-12 21:58:59,127] {manager.py:163} INFO - Launched DagFileProcessorManager with pid: 23340
[2022-01-12 21:58:59,128] {scheduler_job.py:1114} INFO - Resetting orphaned tasks for active dag runs
[2022-01-12 21:58:59,136] {settings.py:52} INFO - Configured default timezone Timezone('UTC')
[2022-01-12 21:59:00 +0800] [23205] [ERROR] Connection in use: ('0.0.0.0', 9999)
[2022-01-12 21:59:00 +0800] [23205] [ERROR] Retrying in 1 second.
[2022-01-12 21:58:59,966] {scheduler_job.py:644} ERROR - Exception when executing SchedulerJob._run_sche
duler_loopTraceback (most recent call last):
  File "/data/home/djy5/deepmd-2.0.0/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1276, 
in _execute_context    self.dialect.do_execute(
  File "/data/home/djy5/deepmd-2.0.0/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 608
, in do_execute    cursor.execute(statement, parameters)
psycopg2.errors.InFailedSqlTransaction: current transaction is aborted, commands ignored until end of tr
ansaction block

[SQL: UPDATE job SET state=%(state)s WHERE job.job_type = %(job_type_1)s AND job.state = %(state_1)s AND
 job.latest_heartbeat < %(latest_heartbeat_1)s][parameters: {'state': <TaskInstanceState.FAILED: 'failed'>, 'job_type_1': 'SchedulerJob', 'state_1': <T
askInstanceState.RUNNING: 'running'>, 'latest_heartbeat_1': datetime.datetime(2022, 1, 12, 13, 58, 29, 964344, tzinfo=Timezone('UTC'))}](Background on this error at: http://sqlalche.me/e/13/2j85)
```
The corresponding Settings for postgresql in airflow.cfg is
```
executor = LocalExecutor
postgresql+psycopg2://airflow:airflow@127.0.0.1:5432/airflow
```
and when i type rollback; in postgresql:
```
postgres=# rollback;
NOTICE:  there is no transaction in progress
ROLLBACK
```
the postgreaql is` postgresql-9.2.24-7.el7_9.x86_64`,and airflow is `airflow 2.2.3` This problem really confused me for several days, somebody help me OUT? thanks

**Comment by potiuk** [comment]
I tihnk this is something in your DAGs. Currently airflow scheduler parses the DAG in the same applications as scheduler. 

This line:

```
[2022-01-12 21:58:59 +0800] [23205] [ERROR] Connection in use: ('0.0.0.0', 9999)
```

Indicates that something in the top-level of one of your DAGs tries to listen on port 9999 and likely there are more than one of those and this eventually when happens too often might cause some problems for scheduler. 

Follow the best practices of your DAG top-level code https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#top-level-python-code

**Reply by changxiaoju** [reply]
thankyou, i don't have dag in airflow/dags/. i used to have one , but i removed it after i switched the database from sql to postgresql, so it is empty now, does this matter?

**Reply by changxiaoju** [reply]
i used to perform `aIrflow scheduler`  without any dag, but now i can not.

**Reply by potiuk** [reply]
This was a guess that it is a DAG. If not a DAG. then  I have no idea what listens to 9999 port - can you please check what part of the system does it ? None of the airlfow components does it - unless you somehow configured airflow, added some plugins or otherwise modifiedf the deployment. 

Sorry - can't help more - you need to see what listens to port 9999 (Glad it's not 666 BTW ;) )

**Reply by potiuk** [reply]
might be some automated instrumentation, debugging harness or smth like that outside of Airflow - but quite for sure it's not Airflow.

**Reply by changxiaoju** [reply]
thankyou, i choose 9999 rather than 8080(as default)  because i guess no one will use such a weird number. 8080 will always rise up with `[ERROR] Connection in use: ('0.0.0.0', 8080)` when performming  `airflow webserver --port 8080`
but 9999or9988, these random ports are always available to use, and i do switch many of them, and they all work fine with `airflow webserver --port "weird numbers"`. so i guess `[ERROR] Connection in use: ('0.0.0.0', 9999)`  is not the real cause of aIrflow scheduler crashes, because when poerforming  `airflow webserver --port 9999`, everything is ok.
thankyou again

**Reply by potiuk** [reply]
Then apparently you hav more than one scheduler and webserver running for some reason and this is your problem likely

**Reply by changxiaoju** [reply]
However, with `ps -elf | grep airflow` no scheduler and webserver is running.

**Reply by omar-machat** [reply]
Hello @changxiaoju, i still have the same issue. 
Did you figure out how to solve it ?

Thank you

**Reply by Hamid-abdellaoui** [reply]
hello @omar-machat  @changxiaoju , I have tha same problem, did find any solution please ?

THANKS

**Reply by potiuk** [reply]
Ok. I thinm what **could be** the reaosn. Apparently you are using unsupported postgres

`postgresql-9.2.24-7.el7_9.x86_64`

Airflow supports postgres 9.6+ in 2.2.* version but since Posgres 9.6 is EOL as of November 11  last year. https://www.lansweeper.com/eol/postgresql-end-of-life/#:~:text=PostgreSQL%209.6%20is%20going%20end,after%20just%20over%205%20years.) the new releases will support only Postgres 10+ (https://github.com/apache/airflow#requirements) 

Upgrade.
